src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,510 @@
+                     Pem.X_509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL, Pem.CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(
+                            pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given {@code String}.
+     *
+     * <p> This method reads the {@code String} until PEM data is found
+     * or the end of the {@code String} is reached.  If no PEM data is found,
+     * an {@code IllegalArgumentException} is thrown.
+     *
+     * <p> This method returns a Java API cryptographic object,
+     * such as a {@code PrivateKey}, if the PEM type is supported.
+     * Any non-PEM data preceding the PEM header is ignored by the decoder.
+     * Otherwise, a {@link PEMRecord} will be returned containing
+     * the type identifier and Base64-encoded data.
+     * Any non-PEM data preceding the PEM header will be stored in
+     * {@code leadingData}.
+     *
+     * <p> Input consumed by this method is read in as
+     * {@link java.nio.charset.StandardCharsets#UTF_8 UTF-8}.
+     *
+     * @param str a String containing PEM data
+     * @return a {@code DEREncodable}
+     * @throws IllegalArgumentException on error in decoding or no PEM data
+     * found
+     * @throws NullPointerException when {@code str} is null
+     */
+    public DEREncodable decode(String str) {
+        Objects.requireNonNull(str);
+        DEREncodable de;
+        try {
+            return decode(new ByteArrayInputStream(
+                str.getBytes(StandardCharsets.UTF_8)));
+        } catch (IOException e) {
+            // With all data contained in the String, there are no IO ops.
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given
+     * {@code InputStream}.
+     *
+     * <p> This method reads from the {@code InputStream} until the end of
+     * the PEM footer or the end of the stream. If an I/O error occurs,
+     * the read position in the stream may become inconsistent.
+     * It is recommended to perform no further decoding operations
+     * on the {@code InputStream}.
```

Comment:
I prefer the words in the previous commit. It was more positive -- the read position is here when there is no IOE.